aspnetcore/performance/performance-best-practices.md

```diff
@@ -132,19 +131,200 @@ Recommendations:
+* LOH is collected with the rest of the heap. LOH requires a full [garbage collection](/dotnet/standard/garbage-collection/fundamentals) or [Gen2 collection](/dotnet/standard/garbage-collection/fundamentals#generations).
+
+This [blog post](https://adamsitnik.com/Array-Pool/#the-problem) describes the problem succinctly:
+
+> When a large object is allocated, it’s marked as Gen 2 object. Not Gen 0 as for small objects. The consequences are that if you run out of memory in LOH, GC cleans up the whole managed heap, not only LOH. So it cleans up Gen 0, Gen 1 and Gen 2 including LOH. This is called full garbage collection and is the most time-consuming garbage collection. For many applications, it can be acceptable. But definitely not for high-performance web servers, where few big memory buffers are needed to handle an average web request (read from a socket, decompress, decode JSON & more).
+
+Naively storing a large request or response body into a single `byte[]` or `string`:
+
+* May result in quickly running out of space in the LOH.
+* May cause performance issues for the app because of full GCs running.
+
+## Use buffered and synchronous reads and writes as an alternative to asynchronous reading and writing
+
+When using a serializer/de-serializer that only supports synchronous reads and writes:
+
+* Choose to buffer the data into memory before passing data into the serializer/de-serializer.
+* [JSON.NET](https://www.newtonsoft.com/json/help/html/Introduction.htm) is only synchronous.
+
+> [!WARNING]
+> If the request is large, it could lead to an out of memory (OOM) condition. OOM can result in a Denial Of Service.  For more information, see [Avoid reading large request bodies or response bodies into memory](#arlb) in this document.
+
+ASP.NET Core 3.0 uses <xref:System.Text.Json> by default for JSON serialization. <xref:System.Text.Json>:
+
+* Reads and writes JSON asynchronously.
+* Is optimized for UTF-8 text.
+* Typically higher performance than `Newtonsoft.Json`.
+
+## Do not store IHttpContextAccessor.HttpContext in a field
+
+The [IHttpContextAccessor.HttpContext](xref:Microsoft.AspNetCore.Http.IHttpContextAccessor.HttpContext) returns the `HttpContext` of the active request when accessed from the request thread. The `IHttpContextAccessor.HttpContext` should **not** be stored in a field or variable.
+
+**Do not do this:** The following example stores the `HttpContext` in a field, and then attempts to use it later.
+
+[!code-csharp[](performance-best-practices/samples/3.0/MyType.cs?name=snippet1)]
+
+The preceding code frequently captures a null or incorrect `HttpContext` in the constructor.
+
+**Do this:** The following example:
+
+* Stores the <xref:Microsoft.AspNetCore.Http.IHttpContextAccessor> in a field.
+* Uses the `HttpContext` field at the correct time and checks for `null`.
+
+[!code-csharp[](performance-best-practices/samples/3.0/MyType.cs?name=snippet2)]
+
+## Do not access HttpContext from multiple threads
```

Comment:
The highlights are messed up in this section.

![image](https://user-images.githubusercontent.com/95136/66107853-1e5fb080-e576-11e9-913d-ba58f6d7459e.png)

Reply:
it's still busted https://review.docs.microsoft.com/en-us/aspnet/core/performance/performance-best-practices?view=aspnetcore-3.1&branch=pr-en-us-14734#do-not-access-httpcontext-from-multiple-threads

Reply:
It's correct on edge/FF and Chrome at 80%. The build team (who writes the HTML generator for all of doc.ms.com) broke it 6 months ago and have only managed to mostly fix it.

Edge:
![image](https://user-images.githubusercontent.com/3605364/66226387-62d27580-e676-11e9-92cf-496c45a33743.png)

FF:
![image](https://user-images.githubusercontent.com/3605364/66226410-71b92800-e676-11e9-922f-d870c6989f47.png)

Chrome at 80%:

![image](https://user-images.githubusercontent.com/3605364/66226460-8bf30600-e676-11e9-8909-de32958d8f43.png)

Reply:
@davidfowl  see #14936
I asked for the problem to be  prioritized on the internal bug.